File: medium/396-rotate-function/solution.py
from typing import List


class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        """
        396. Rotate Function
        Time: O(n)
        Space: O(1)
        """
        n = len(nums)
        total = sum(nums)
        current = sum(i * num for i, num in enumerate(nums))
        best = current

        for k in range(1, n):
            current += total - n * nums[n - k]
            best = max(best, current)

        return best

File: medium/396-rotate-function/solution.java
/**
 * 396. Rotate Function
 * Time: O(n)
 * Space: O(1)
 */
class Solution {
    public int maxRotateFunction(int[] nums) {
        int n = nums.length;
        long total = 0;
        long current = 0;

        for (int i = 0; i < n; i++) {
            total += nums[i];
            current += (long) i * nums[i];
        }

        long best = current;

        for (int k = 1; k < n; k++) {
            current += total - (long) n * nums[n - k];
            best = Math.max(best, current);
        }

        return (int) best;
    }
}
